[TextureMapper] preserve-3d causes unnecessary layer splits

zhani opened the pull request:
#### b54433896d895f24684d536e069c1f848137993a
<pre>
[TextureMapper] preserve-3d causes unnecessary layer splits
<a href="https://bugs.webkit.org/show_bug.cgi?id=284026">https://bugs.webkit.org/show_bug.cgi?id=284026</a>

Reviewed by Fujii Hironori.

The BSP tree calculation currently includes layers that have dimensions
but produce no visible output. This results in unnecessary splits during
the preserve-3D process. Such layers should be excluded from the BSP tree
calculations to optimize rendering and reduce computational overhead.

* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
 (WebCore::TextureMapperLayer::paintWith3DRenderingContext):
 (WebCore::TextureMapperLayer::collect3DRenderingContextLayers):
 (WebCore::TextureMapperLayer::collect3DSceneLayers): Deleted.
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:

Canonical link: <a href="https://commits.webkit.org/287417@main">https://commits.webkit.org/287417@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c05fab43b68abfd38fc897e35734c92898154560

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79521 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58530 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32898 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84100 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30614 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81655 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67623 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6809 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62185 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20048 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82592 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52240 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72450 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42495 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49588 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26613 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29030 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70707 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27071 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85511 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6787 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4727 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70434 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6952 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68312 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69678 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13705 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12603 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12296 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6739 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/12440 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6635 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10117 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8434 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -1070,20 +1070,26 @@ void TextureMapperLayer::paintWith3DRenderingContext(TextureMapperPaintOptions&
+            || hasFilters() || hasMask() || hasBackdrop();
+
+        // Add layers to 3d rendering context only if they get actually painted.
+        if (m_state.visible && (hasVisualContent || (isLeafOf3DRenderingContext && m_children.size() > 0)))
```
I think `(isLeafOf3DRenderingContext && m_children.size() > 0)` can be replaced with `isFlattened()`. Right?

zhani replied:
No, the specification states: "The element establishing the 3D rendering context, and each other 3D transformed element participating in the 3D rendering context, is rendered into its own plane." This means that if a leaf element has no 3D transformation, it should not be flattened. TextureMapper follows this logic.

Additionally, the condition `(isLeafOf3DRenderingContext && m_children.size() > 0)` addresses cases where `RenderLayerBacking` inserts a zero-sized, non-3D transformed layer into a leaf node. This occurs in `RenderLayerBacking::updateTransformFlatteningLayer`. This layer usually has visible children. Also this condition is necessary for situations like iframe leaf nodes. These cases are on my to-do list for further investigation. However, I think it is more critical to address the performance regression with the posterl-circles first.

fujii replied:
Got it. `m_children.size() > 0)` should be `!m_children.isEmpty()`.

zhani replied:
Fixed

zhani committed:
[TextureMapper] preserve-3d causes unnecessary layer splits
https://bugs.webkit.org/show_bug.cgi?id=284026

Reviewed by Fujii Hironori.

The BSP tree calculation currently includes layers that have dimensions
but produce no visible output. This results in unnecessary splits during
the preserve-3D process. Such layers should be excluded from the BSP tree
calculations to optimize rendering and reduce computational overhead.

* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
 (WebCore::TextureMapperLayer::paintWith3DRenderingContext):
 (WebCore::TextureMapperLayer::collect3DRenderingContextLayers):
 (WebCore::TextureMapperLayer::collect3DSceneLayers): Deleted.
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:

Canonical link: https://commits.webkit.org/287417@main
